iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/helpers/Tools.java

```diff
@@ -139,6 +153,98 @@ public static Device addDeviceWithRetry(RegistryManager registryManager, Device
         return ret;
     }
 
+    public static void addDevicesWithRetry(List<Device> devices, String connectionString) throws IotHubException, IOException, InterruptedException
+    {
+        // IoT hub only allows for bulk adding of devices at up to 100 per request, so take the provided devices iterable
+        // and break it into 100 devices or smaller chunks
+        List<Device> subIterable = new ArrayList<>();
+        List<Device> devicesClone = new ArrayList<>(); //create a clone of the source list so elements can be removed from it instead
+        devicesClone.addAll(devices);
+        while (devicesClone.size() > 0)
+        {
+            Device device = devicesClone.remove(0);
+            subIterable.add(device);
+
+            // wait until sub list has the Iot Hub limit of 100 devices to add, or until there will be no more devices to add
+            if (subIterable.size() > 99 || devicesClone.size() <= 0)
```

Comment:
Should this be if or while? Are we not adding only one to the list or am I missing something?

Reply:
This if block represents what should happen once a list of 100 devices is ready to be added, or if a list of under 100 devices is ready to add and there are no more devices to add after this list

Reply:
The purpose here is so that this method can take a list of arbitrary length, and chunk that list into smaller lists of 100 devices or fewer so that we can use the bulk add service API which has a limit of 100 devices per request